pkg/scd/operational_intents_handler.go

```diff
@@ -16,6 +16,42 @@ import (
 	"github.com/interuss/stacktrace"
 )
 
+// subscriptionCanBeRemoved will check if:
+// - a previous subscription was attached,
+// - if so, if it was an implicit subscription
+// - if so, if we can remove it after creating the new implicit subscription
+//
+// This is to be used in contexts where an implicit subscription may need to be cleaned up.
+// NOTE: this should eventually be pushed down to CRDB as part of the queries being executed in the callers of this method.
+//
+//	See https://github.com/interuss/dss/issues/1059 for more details
+func subscriptionCanBeRemoved(ctx context.Context, r repos.Repository, subscriptionID *dssmodels.ID) (bool, error) {
+	// Get the Subscription supporting the OperationalIntent, if one is defined
+	if subscriptionID != nil {
+		sub, err := r.GetSubscription(ctx, *subscriptionID)
+		if err != nil {
+			return false, stacktrace.Propagate(err, "Unable to get OperationalIntent's Subscription from repo")
+		}
+		if sub == nil {
+			return false, stacktrace.NewError("OperationalIntent's Subscription missing from repo")
+		}
+
+		if sub.ImplicitSubscription {
+			// Get the Subscription's dependent OperationalIntents
+			dependentOps, err := r.GetDependentOperationalIntents(ctx, sub.ID)
+			if err != nil {
+				return false, stacktrace.Propagate(err, "Could not find dependent OperationalIntents")
+			}
+			if len(dependentOps) == 0 {
+				return false, stacktrace.NewError("An implicit Subscription had no dependent OperationalIntents")
+			} else if len(dependentOps) == 1 {
+				return true, nil
```

Comment:
We're assuming the one dependent operational intent reference is the one the caller is interested in, but that assumption is neither documented nor enforced.  If we're just checking a particular subscription, then the answer shouldn't be "yes" or "no", it should be "only when removing this particular operational intent reference" (returning an optionally-nil pointer to an operational intent ID).  Or, if we're checking whether a particular subscription can be removed after removing the dependency of an operational intent reference on that subscription, then the ID of the operational intent should be passed into this function and validated that the single dependent operational intent is the operational intent passed.

The operational intent ID was not needed previously because the context in which the code used to be located had just identified this subscription as the one the operational intent was using.